### PR TITLE
Add Bitrev permutation in CUDA

### PR DESCRIPTION
**Pull request by matias-gonz**

### CODE REVIEW

**Review by MegaRedHand — APPROVED** [review]

File: gpu/src/cuda/shaders/bitrev_permutation.cu [hunk 1]
```diff
@@ -0,0 +1,15 @@
+#include "./fp_u256.cuh"
+#include "./utils.h"
+
+extern "C"
+{
+    __global__ void bitrev_permutation(
+        const p256::Fp *input,
+        p256::Fp *result
+    ) {
+        unsigned index = blockIdx.x * blockDim.x + threadIdx.x;
```
Comment (xqft) [inline gpu/src/cuda/shaders/bitrev_permutation.cu:10]:
```suggestion
        unsigned index = threadIdx.x;
```
Same as https://github.com/lambdaclass/lambdaworks/pull/310#discussion_r1187481923

**Review by xqft — APPROVED** [review]
Excellent!

**Review by mpaulucci — APPROVED** [review]

File: gpu/src/cuda/abstractions/state.rs [hunk 2]
```diff
@@ -244,3 +274,67 @@ impl<F: IsField> CalcTwiddlesFunction<F> {
...
+        let config = LaunchConfig {
+            grid_dim,
+            block_dim,
+            shared_mem_bytes: 0,
+        };
+        unsafe {
+            self.function
+                .clone()
+                .launch(config, (&mut self.input, &self.result))
+        }
```
Comment (Oppen) [inline gpu/src/cuda/abstractions/state.rs:325]: Why is this unsafe? Could you add a comment explaining both why the call is unsafe and why we're doing the right thing and not making things unsound?

File: gpu/src/cuda/fft/ops.rs [hunk 3]
```diff
@@ -51,23 +50,16 @@ pub fn gen_twiddles<F: IsFFTField>(
...
-        if bit_reversed_index > i {
-            input.swap(i, bit_reversed_index);
-        }
-    }
-}
+pub fn bitrev_permutation<F: IsFFTField>(
+    input: Vec<FieldElement<F>>,
+    state: &CudaState,
+) -> Result<Vec<FieldElement<F>>, CudaError> {
+    let result = input.clone();
```
Comment (Oppen) [inline gpu/src/cuda/fft/ops.rs:57]: You could pass the same reference twice instead of cloning here. They are not mutable so it should be allowed.

**Pull request merged**